CIAOS-Rust/server/src/main.rs

```diff
@@ -0,0 +1,267 @@
+    }
+    
+    info!("Starting chunk load");
+    let mut bytes = BytesMut::new();
+    while let Some(chunk) = payload.next().await {
+        let chunk = chunk.map_err(actix_web::error::ErrorInternalServerError)?;
+        bytes.extend_from_slice(&chunk);
+    }
+
+    if bytes.is_empty() {
+        error!("No data uploaded with key: {}", key);
+        return Ok(HttpResponse::BadRequest().body("No data was uploaded"));
+    }
+    
+    info!("Total received data size: {} bytes", bytes.len());
+    
+    //sends the bytes to storage.rs so that the flatbuffer can deserialise it and write to bin and than return offset and size vec list
+    let offset_size_list = storage::write_files_to_storage(&bytes)?;
+
+
+    if offset_size_list.is_empty()  {
+        error!("No data in data list with key: {}", key);
+        return Ok(HttpResponse::BadRequest().body("No data in data list"));
+    }
+    
+    info!("Serializing offset and size and uploading");
+
+    let offset_size_bytes = serialize_offset_size(&offset_size_list)?;
+    
+    upload_sql(&key, &offset_size_bytes)
+        .map_err(actix_web::error::ErrorInternalServerError)?;
+
+    info!("Data uploaded successfully with key: {}", key);
+    Ok(HttpResponse::Ok().body(format!("Data uploaded successfully: key = {}", key)))
+}
+
+
+#[actix_web::get("/get/{key}")]
+async fn retrieve_data(key: web::Path<String>) -> Result<HttpResponse, Error> {
+    let key = key.into_inner();
+    info!("checking key  : {}", key);
+
+    if !check_key(&key).map_err(actix_web::error::ErrorInternalServerError)? {
+        warn!("Key does not exist: {}", key);
+        return Ok(HttpResponse::NotFound().body(format!("No data found for key: {}", key)));
+    }
+    info!("Retrieving data for key: {}", key);
+
+    // Connect to the SQLite database and retrieve offset and size data
+    let offset_size_bytes = match get_offset_size_lists(&key) {
+        Ok(offset_size_bytes) => offset_size_bytes,
+        Err(e) => {
+            warn!("Key does not exist or database error: {}", e);
+            return Ok(HttpResponse::BadRequest().body("Key does not exist"));
+        }
+    };
+
+    // Deserialize offset and size data
+    let offset_size_list = deserialize_offset_size(&offset_size_bytes)?;
+
+    let data = get_files_from_storage(offset_size_list)?;
+
+    // Return the FlatBuffers serialized data
+    Ok(HttpResponse::Ok()
+        .content_type("application/octet-stream")
+        .body(data))
+}
+
+#[actix_web::post("/append/{key}")]
+async fn append_files(
+    key: web::Path<String>,
+    mut payload: web::Payload,
+) -> Result<HttpResponse, Error> {
+    let key = key.into_inner();
+    info!("appending data with key: {}", key);
+    if !check_key(&key).map_err(actix_web::error::ErrorInternalServerError)? {
+        warn!("Key does not exist: {}", key);
+        return Ok(HttpResponse::NotFound().body(format!("No data found for key: {}", key)));
+    }
+    info!("Starting chunk load");
```

Comment:
Looks like all these preprocessing is common for all api's can we move it to a function?

Reply:
hey so can you exactly refer to which part?
the logs?